### ISSUE TITLE
Automatic Keyboard Focus

### ISSUE THREAD

**Issue by Mjolinir** [asbru-cm/asbru-cm#148]
**Is your feature request related to a problem? Please describe.**
When clicking out of the Asbru window into other windows, and then back again, you can no longer type in any session tab that you are connected to unless you click the "Set keyboard focus" button.

**Describe the solution you'd like**
When you click back into Asbru, make the session tab you click on (or the one on top) to become the focused window automatically. You can then get rid of that extremely wide button and regain valuable remote session space.

**Describe alternatives you've considered**
I have not been able to find any work around for this issue. It is extremely annoying, I find my windows constantly flip-flopping when I click back over and start typing, only to see crazyness happening because Asbru didnt really regain focus, so the WM starts interpreting my keystrokes as keybinds.

**Comment by gfrenoy** [comment]
Sorry but I don't understand that one and do not experience the same on my environment.

Which "Set keyboard focus" button are your referring to ?  Can you maybe show a few screenshots or screencast so we can better visualize what is happening ?

Another question: is this is a "new issue" ?  Note also that a change has happened recently (see [here](https://github.com/asbru-cm/asbru-cm/commit/51203152efdfce826e85622f122a563b4b6656c6)) so it might be related...

**Comment by Mjolinir** [comment]
Sure, the button is in the image attached. For me, this issue has always existed. Perhaps there is a way to set the focus automatically, but I have never found it. 

![Screenshot_20190520_081824](https://user-images.githubusercontent.com/7074712/58021085-ffd1af00-7ad7-11e9-884b-8384d13e513d.png)

**Comment by gfrenoy** [comment]
Ah ok, it's a RDP connection :)  Now, I understand this a little bit better.

But still, I'm not able to reproduce it.  I need to find a Windows machine to test this further...

**Comment by Mjolinir** [comment]
It happens to me for ssh sessions as well. While working in some other
window, clicking back over to an ssh session in Asbru and start
typing....and either the text Im typing appears in Asbru's search window,
or it get interpreted by the windows manager as key commands and weird
stuff starts happening, etc, even when I click on the session I want to
type in first. I have co-workers who have the same issues.

**Comment by gfrenoy** [comment]
For SSH, I can test and better understand what is happening and actually it looks it works like this : if the focus was on the connection tree before you lost the focus, it will be set back to the connection tree when you get it back.  If you start typing, it will initiate a search.  Not sure there's something wrong with this...?

**Comment by Mjolinir** [comment]
OK, I understand that. To me, and I would imagine most users, the search function is almost never used, so that functionality is not very useful. I would think the focus should be always placed back into the tab of the open session on top, does that make sense?

**Comment by gfrenoy** [comment]
Well, I don't pretend being "most users" but I do use the search feature every day :)  When you have nearly a hundred entries in your list, it is way more efficient to find what you are looking for.

Anyway, this is not the point.  I think the root question is : why the focus was on the treeview when Ásbrú lost the focus ?  If you don't use it, no reason to be there...

And that being said, to make sure everybody is happy, we can also add an option "Always set focus on the active tab" (or something similar to make it clear)...

**Comment by Mjolinir** [comment]
I think that option would be awesome if it existed :)

**Comment by brunt82** [comment]
I had the same experience than Mjolinir: As soon as I switch from a windows to a debian session, I'm unable to use the keyboard of the Windows session anymore.

Until today I closed the session and reconnect to get the focus again. I never realised the button "set keyboard focus". 

Anyhow: for Debian sessions this problem does not appear. There it is always possible to use the keyboard as soon as I switched to the tab. Therefore it would be nice, if the focus same happens for windows sessions.

**Comment by Mjolinir** [comment]
SO I was incorrect, or it has already been fixed, but thie issue i described above doesnt occur on ssh sessions. I think some work needs to be done on the RDP side. In addition to the issue I described above, and also depending on the distribution (Fedora), RDP sessions do not always size them self to fit the window, and many times you have to scroll up and down or left and right to get to the Start button or close a Window

**Comment by gfrenoy** [comment]
> RDP sessions do not always size them self to fit the window

Thanks to @figue (see #153) ; it has been fixed and it's available in the latest snapshot.  Please note you'll need to install an additional package (X11::GuiTest).

**Comment by Mjolinir** [comment]
So as of version 5.2.1 on Fedora 30, the RDP issue with the keyboard focus does not seem to be resolved. I still have to click that "Set keyboard focus" button every time before I can actually type anything in the RDP session. If you guys added the "Always set focus on the active tab"  or something similar, I'm not seeing it, where would that menu option be placed?

**Comment by gfrenoy** [comment]
Only the sizing issue has been fixed.  Nothing has been done about the focus issue.

**Comment by figue** [comment]
This issue still exist in GTK3 branch. You have to tap in "Set keyboard focus" when you have more that one RDP tab opened.

**Comment by Mjolinir** [comment]
Yes, this "Set keyboard focus" thing still exists, but the devs only use ssh, so they don't seem interested in fixing it. For those of us that have to deal with Windows and RDP, this issue is incredibly inconvenient

**Comment by gfrenoy** [comment]
> devs only use ssh, so they don't seem interested in fixing it

Let me rephrase this to be more correct : devs only use ssh, so they don't have an easy way to correctly understand and reproduce the problem.

See my comment of May 20, believe or not but I personally don't have any Windows machine to test this (and don't want to have one) so if you have one you can share, that would help to have a working environment to support that kind of environment better.

**Comment by figue** [comment]
I only have to connect to Windows at work fortunately...

The point is Ásbrú loose keyboard focus when you have more that one RDP tab. You have to push at focus button every time you change the tab. The idea could be launching same function as "Set keyboard focus" do when you switch to a RDP tab.

I tried some time ago to fix, but I can't myself...

**Comment by Mjolinir** [comment]
Same for me, I am required to deal with Windows & RDP at work. I have no idea how I could share a Windows machine for your testing. You could set one up for free for a limited time in AWS (https://aws.amazon.com/premiumsupport/knowledge-center/free-tier-windows-instance/) or Azure (https://azure.microsoft.com/en-us/free/). Then just open port 3395 on your cloud service to the IP of your Asbru test machine. Open up an RDP session to your Windows test machine, click out of the Windows instance into something else, then click back over to the Windows instance and try to type. It will not work, you have to click that "Set Keyboard Focus" button first. I am no perl programmer, but it seems like the click into the instance should automatically set the keyboard focus without having to click any buttons first.

**Comment by gfrenoy** [comment]
Good suggestion @Mjolinir ; I personally won't create an account for that but if one is OK with Microsoft or Amazon terms and conditions, he/she can maybe create a machine for the project ?  Thanks.

**Comment by Mjolinir** [comment]
Unfortunately, I (he) cannot create anymore, I have already utilized all my allotments on the free tier(s). Perhaps someone else out there has this option open to them?

**Comment by figue** [comment]
If you have Virtualbox, you can enable remote screen RDP with any virtual machine. It's just a workaround...

**Comment by andr3scrivener** [comment]
I'm having the same problem, the strange thing is that on my older notebook it works perfectly, but on the new one I have this keyboard focus problem.

Can I help in any way for solution? Any dependencies or logs command?

**Comment by gfrenoy** [comment]
@andr3scrivener see my previous comment(s), if you can provide the project with some testing environment, that would certainly help us reproducing / understanding (and hopefully fixing ;p) the root issue.

That being repeated, @figue's comment put me on a good track, thanks for that !

I don't use VirtualBox but I remembered Microsoft was offering temporary VM to test their browsers (see [here](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/) if you are interested).  It works fine :)  So I now have a Win10 virtual machine on which I can enable Remote Deskop and test this is better , yeah !

**Comment by figue** [comment]
Even if you use a Linux virtual machine, you can connect to console by RDP. Virtualbox offers that option:
![vbox](https://user-images.githubusercontent.com/3238330/68670751-eb81e280-054d-11ea-9d44-d94d50a54f28.png)
You have just connect to localhost:3389

**Comment by gfrenoy** [comment]
Thanks @figue ; I remember that feature, that's indeed a good idea but I don't use VirtualBox anymore.  I'm all set with the Microsoft test program and got a Win10 virtual machine to test now.  Thanks for your hint, it really helped!

So from my investigations to far, the bottom line is that I was not able to find a signal that would let me know when the RDP window (that is an "external" XWindow) has been clicked.

If anyone has an idea on how to do this, it will be greatly appreciated !

Still, I made a [small change](https://github.com/asbru-cm/asbru-cm/commit/52f4fd3e49974034a9227a73b6d3f7695a5493ab) though that is improving a little bit the situation ; let me know if this helps (only for the gtk3 branch).

**Comment by figue** [comment]
Quick test seems good... But let me try tomorrow at work (I can only make a small test with a Virtualbox and a server in Azure right now)... Thanks!

**Comment by figue** [comment]
After your commit seems to be fixed. Thanks!

**Comment by gfrenoy** [comment]
> After your commit seems to be fixed. Thanks!

Looks like a good news :)  Thanks for your feedback.

@Mjolinir  @andr3scrivener How does it look on your side ?

**Comment by Mjolinir** [comment]
I'm not sure your commit build has trickled it way into the Fedora repos yet, I haven't saw any updates to the package yet.

**Comment by figue** [comment]
You don't need to install @Mjolinir just clone repo, change branch to gtk3 and execute:
```
git clone https://github.com/asbru-cm/asbru-cm.git
cd asbru-cm
git checkout gtk3
./asbru-cm
```
If you want to update, just execute "git pull".

Cheers

**Comment by gfrenoy** [comment]
Packages are built automagically (thanks again Klaas ;) but you'll need to use the gtk3 (6.0.0) version...

**Comment by Mjolinir** [comment]
So I am not sure which branch I am using, I just did a 'dnf install asbru-cm' whenever I migrated over from PAC Manager. I looked at your (@gfrenoy ) commit, and attempted to apply the changes manually. The code for me was slightly different - the 5 removed lines were 3 lines in the file on my system. I removed those 3 lines and added the new line, and my first tests have been positive! 

So what is the difference between the gtk3 branch and others? Should I be using that one instead?

I will run this manually patched version today and see how it goes and report back, but so far so good. :)

**Comment by andr3scrivener** [comment]
> > After your commit seems to be fixed. Thanks!
> 
> Looks like a good news :) Thanks for your feedback.
> 
> @Mjolinir @andr3scrivener How does it look on your side ?

Hi guys, 

thanks, working correctly.



PS: If anyone has the following issues below in dist Fedora

**Problem:**

```
[asbru-cm]$ ./asbru-cm
Can't locate Gtk3.pm in @INC (you may need to install the Gtk3 module) (@INC contains: /tmp/asbru-cm/lib /tmp/asbru-cm/lib/ex /tmp/asbru-cm/lib/edit /usr/local/lib64/perl5/5.30 /usr/local/share/perl5/5.30 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5) at /tmp/asbru-cm/lib/PACMain.pm line 50.
BEGIN failed--compilation aborted at /tmp/asbru-cm/lib/PACMain.pm line 50.
Compilation failed in require at ./asbru-cm line 88.
BEGIN failed--compilation aborted at ./asbru-cm line 88.
```


**Solution:**

`dnf install perl-Gtk3`


**Problem:**
```
[asbru-cm]$ ./asbru-cm
Can't locate Gtk3/SimpleList.pm in @INC (you may need to install the Gtk3::SimpleList module) (@INC contains: /tmp/asbru-cm/lib /tmp/asbru-cm/lib/ex /tmp/asbru-cm/lib/edit /usr/local/lib64/perl5/5.30 /usr/local/share/perl5/5.30 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5) at /tmp/asbru-cm/lib/PACTerminal.pm line 48.
BEGIN failed--compilation aborted at /tmp/asbru-cm/lib/PACTerminal.pm line 48.
Compilation failed in require at /tmp/asbru-cm/lib/PACMain.pm line 61.
BEGIN failed--compilation aborted at /tmp/asbru-cm/lib/PACMain.pm line 61.
Compilation failed in require at ./asbru-cm line 88.
BEGIN failed--compilation aborted at ./asbru-cm line 88.
```

**Solution:** 

`perl -MCPAN -e "install Gtk3::SimpleList"`

**Comment by Mjolinir** [comment]
Based on @andr3scrivener post above, I an definitely not running the gtk3 branch. Does that mean the default install is using gtk2 libraries instead? Forgive my ignorance here, but why is gtk3 not the default branch, is it just not as fully tested as the original branch? Is there a way to use the gtk3 branch via 'dnf install asbru-cm-gtk3" or something like that, or only by installing it from git as @figue  pointed out earlier?

**Comment by gfrenoy** [comment]
>  Does that mean the default install is using gtk2 libraries instead?

Yes, the current Ásbrú version (5.2.1) is based on Gtk2 that is superseded since more than 5 years now.  Thanks to the great contribution of Egmont and Hans, we are about to have a new Ásbrú version (6.0.0) based on Gtk3.

> why is gtk3 not the default branch

Because we are still in a test phase.  We did not want to rush things too quickly before we are confident it will work on most of the environments.

For this, we'll need help :-)

>  Is there a way to use the gtk3 branch via 'dnf install asbru-cm-gtk3" or something like that,

To use that new version, please go to the gtk3 brach --> https://github.com/asbru-cm/asbru-cm/tree/gtk3 and follow the instructions under "Testing the gtk3 branch"; that are for Fedora:

```
$ curl -s https://packagecloud.io/install/repositories/asbru-cm/gtk3/script.rpm.sh | sudo bash
$ sudo dnf install asbru-cm
```

We are still polishing the last details so you might encounter a few glitches so don't do this if Ásbrú is really critical for you !  That being said, I am using it on a daily basis without any major problem so I'm confident it is stable enough...

Any feedback you can give will be a step forward to an official "6.0.0" release :)

**Comment by gfrenoy** [comment]
> Can't locate Gtk3.pm
> Can't locate Gtk3/SimpleList.pm

@andr3scrivener Thanks for the tip :)  That being said, if you are using our packagecloud.io repository, it should have been installed automagically...

See our [spec file](https://github.com/asbru-cm/asbru-cm/blob/gtk3/dist/rpm/asbru.spec).

**Comment by Mjolinir** [comment]
@gfrenoy Thanks for the info! I would definitely like to help out testing and contributing in any way I can. I will start using the gtk3 branch. Thank you Egmont and Hans for your gtk3 erfforts, and of course @andr3scrivener and @figue as well for your helpful info as well. So far, the manual fix seems to have solved the issue (in the gtk2 branch) MOST of the time. I only had to click that button 2 times today, instead of 1000, so that's a huge improvement!

**Comment by Mjolinir** [comment]
I attempted to switch to gtk3 branch, but seems it may be having an issue?

`curl -s https://packagecloud.io/install/repositories/asbru-cm/gtk3/script.rpm.sh | sudo bash
Detected operating system as fedora/31.
Checking for curl...
Detected curl...
Downloading repository file: https://packagecloud.io/install/repositories/asbru-cm/gtk3/config_file.repo?os=fedora&dist=31&source=script
done.
Installing pygpgme to verify GPG signatures...
main: line 100: yum: command not found
Installing yum-utils...
main: line 115: yum: command not found

WARNING: 
The yum-utils package could not be installed. This means you may not be able to install source RPMs or use other yum features.

Generating yum cache for asbru-cm_gtk3...
main: line 125: yum: command not found
Generating yum cache for asbru-cm_gtk3-source...
main: line 128: yum: command not found

The repository is setup! You can now install packages.
sudo dnf install asbru-cm                                                                   
asbru-cm_gtk3                                                                                                   488  B/s | 833  B  00:01    
asbru-cm_gtk3-source                                                                                       601  B/s | 833  B  00:01    
No match for argument: asbru-cm
Error: Unable to find a match: asbru-cm
`

**Comment by Mjolinir** [comment]
> > Does that mean the default install is using gtk2 libraries instead?
> 
> Yes, the current Ásbrú version (5.2.1) is based on Gtk2 that is superseded since more than 5 years now. Thanks to the great contribution of Egmont and Hans, we are about to have a new Ásbrú version (6.0.0) based on Gtk3.
> 
> > why is gtk3 not the default branch
> 
> Because we are still in a test phase. We did not want to rush things too quickly before we are confident it will work on most of the environments.
> 
> For this, we'll need help :-)
> 
> > Is there a way to use the gtk3 branch via 'dnf install asbru-cm-gtk3" or something like that,
> 
> To use that new version, please go to the gtk3 brach --> https://github.com/asbru-cm/asbru-cm/tree/gtk3 and follow the instructions under "Testing the gtk3 branch"; that are for Fedora:
> 
> ```
> $ curl -s https://packagecloud.io/install/repositories/asbru-cm/gtk3/script.rpm.sh | sudo bash
> $ sudo dnf install asbru-cm
> ```
> 
> We are still polishing the last details so you might encounter a few glitches so don't do this if Ásbrú is really critical for you ! That being said, I am using it on a daily basis without any major problem so I'm confident it is stable enough...
> 
> Any feedback you can give will be a step forward to an official "6.0.0" release :)

**Comment by gfrenoy** [comment]
> main: line 100: yum: command not found

You don't have ```yum``` installed ?  I'm not a Fedora user but I thought it is part of the standard tools, isn't it ?

> Detected operating system as fedora/31.

Ásbrú is not yet packaged for ```fedora/31```.  That machine is not yet available in the packpack project (see https://hub.docker.com/r/packpack/packpack/tags/).

**Comment by Mjolinir** [comment]
ok, got it. 'yum' has been depreciated in favor of 'dnf' as the package tool in RedHat/Fedora. 

Ill just stick with @figue method of testing the gtk3 branch for now. Is there a different place to report issues with that branch, or right here in "Issues"?

**Comment by B1tl0rd** [comment]
yum can still be used, really when you use yum is a symbolic link to dnf

**Comment by gfrenoy** [comment]
>  Is there a different place to report issues with that branch, or right here in "Issues"?

In [Issues](https://github.com/asbru-cm/asbru-cm/issues), yes.  As it becomes stable, we'll likely merge both branches soon, anyway.

In the meantime, let's consider this "keyboard focus" issue closed.  Feel free to reopen if required.

Thanks all for your help.
